Add a new concept page for Sidecar containers (#43346)
kubernetes/website: content/en/docs/concepts/workloads/pods/sidecar-containers.md

Reviewer: Either in this PR or a follow up, let's change the diagram in https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-lifetime to show the containers in a Pod using something like a Gantt chart.

---
title: Sidecar Containers
content_type: concept
weight: 50
---

<!-- overview -->
{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

Sidecar containers are the secondary containers that run along with the main
application container within the same {{< glossary_tooltip text="Pod" term_id="pod" >}}.
These containers are used to enhance or to extend the functionality of the main application
container by providing additional services, or functionality such as logging, monitoring,
security, or data synchronization, without directly altering the primary application code.

<!-- body -->

> 

Reviewer: ```suggestion
## Enabling sidecar containers
```

Author: Updated

Starting with Kubernetes 1.28, a feature gate named `SidecarContainers` allows you to
specify a `restartPolicy` for containers listed in a Pod's `initContainers` field. These
restartable _sidecar_ containers are independent with other [init containers](/docs/concepts/workloads/pods/init-containers/) and main application container within
the same pod. These can be started, stopped, or restarted without effecting the main
application container and other init containers.

> 

Reviewer: ```suggestion

## Sidecar containers and Pod lifecycle

```

Author: Updated.

specify a `restartPolicy` for containers listed in a Pod's `initContainers` field. These
restartable _sidecar_ containers are independent with other [init containers](/docs/concepts/workloads/pods/init-containers/) and main application container within
the same pod. These can be started, stopped, or restarted without effecting the main
application container and other init containers.

If an init container is created with its `restartPolicy` set to `Always`, it will
start and remain running during the entire life of the Pod. This can be helpful for
running supporting services separated from the main application containers.

If a `readinessProbe` is specified for this init container, its result will be used
to determine the `ready` state of the Pod.

Since these containers are defined as init containers, they benefit from the same
ordering and sequential guarantees as other init containers, allowing them to
be mixed with other init containers into complex Pod initialization flows.

Compared to regular init containers, sidecar-style init containers continue to
run and the next init container can begin starting once the kubelet has set

> the `started` container status for the sidecar-style init container to true.

Author: Updated.

That status either becomes true because there is a process running in the
container and no startup probe defined, or
as a result of its `startupProbe` succeeding.

> This feature can be used to implement the sidecar container pattern in a more
> robust way, as the kubelet always restarts a sidecar container if it fails.

Reviewer: ```suggestion
```

It's actually OK to omit this here.

Here's an example of a Deployment with two containers, one of which is a sidecar:

{{% code_sample language="yaml" file="application/deployment-sidecar.yaml" %}}

This feature is also useful for running Jobs with sidecars, as the sidecar
container will not prevent the Job from completing after the main container
has finished.

Here's an example of a Job with two containers, one of which is a sidecar:

{{% code_sample language="yaml" file="application/job/job-sidecar.yaml" %}}
By default, this feature is not available in Kubernetes. To avail this feature, you

> need to enable the feature gate named `SidecarContainers` by setting the value **true** to it.

Author: Updated

sidecar containers support [probes](/docs/concepts/workloads/pods/pod-lifecycle/#types-of-probe) to control their lifecycle.

These containers can interact directly with the main application containers, sharing
the same network namespace, filesystem, and environment variables. They work closely
together to provide additional functionality.

## Resource sharing within containers

Given the ordering and execution for init containers, the following rules
for resource usage apply:

* The highest of any particular resource request or limit defined on all init
  containers is the *effective init request/limit*. If any resource has no
  resource limit specified this is considered as the highest limit.
* The Pod's *effective request/limit* for a resource is the higher of:
  * the sum of all app containers request/limit for a resource
  * the effective init request/limit for a resource
* Scheduling is done based on effective requests/limits, which means
  init containers can reserve resources for initialization that are not used
  during the life of the Pod.
* The QoS (quality of service) tier of the Pod's *effective QoS tier* is the

>   QoS tier for init containers and app containers alike.

Reviewer: This ought to mention sidecars as well.

Author: @sftim,
Can you please help me to find information on it.

Reviewer: What information would you like? If you want to know how the feature works, you might want to ask SIG Node (I'd use Slack for that conversation).

Reviewer: You might be able to work this out yourself thinking about what the resource use might be at each stage of a Pod's lifecycle.

Quota and limits are applied based on the effective Pod request and
limit.

Pod level control groups (cgroups) are based on the effective Pod request and
limit, the same as the scheduler.

> {{< note >}}
> This section also present under [init containers](/docs/concepts/workloads/pods/init-containers/) page.
> {{< /note >}}

Reviewer: ```suggestion
```

Author: Updated.

and restarted independently of regular containers. This means you can update, scale, or
maintain sidecar containers without affecting the primary application.

Sidecar containers share the same network and storage namespaces with the primary
container This co-location allows them to interact closely and share resources.

## Differences from init containers

Sidecar containers work alongside the main container, extending its functionality and
providing additional services.

Sidecar containers run concurrently with the main application container. They are active
throughout the lifecycle of the pod and can be started and stopped independently of the
main container. Unlike [init containers](/docs/concepts/workloads/pods/init-containers/),
sidecar containers support [probes](/docs/concepts/workloads/pods/pod-lifecycle/#types-of-probe) to control their lifecycle.

These containers can interact directly with the main application containers, sharing
the same network namespace, filesystem, and environment variables. They work closely
together to provide additional functionality.

## Resource sharing within containers

> 

Reviewer: ```suggestion

{{< comment >}}
This section is also present in the [init containers](/docs/concepts/workloads/pods/init-containers/) page.
If you're editing this section, change both places.
{{< /comment >}}

```

Author: Updated.

> Given the ordering and execution for init containers, the following rules
> for resource usage apply:

Reviewer: ```suggestion
Given the order of execution for init, sidecar and app containers, the following rules
for resource usage apply:
```

Author: Updated.

the same network namespace, filesystem, and environment variables. They work closely
together to provide additional functionality.

## Resource sharing within containers

{{< comment >}}
This section is also present in the [init containers](/docs/concepts/workloads/pods/init-containers/) page.
If you're editing this section, change both places.
{{< /comment >}}

Given the order of execution for init, sidecar and app containers, the following rules
for resource usage apply:

* The highest of any particular resource request or limit defined on all init
  containers is the *effective init request/limit*. If any resource has no
  resource limit specified this is considered as the highest limit.
* The Pod's *effective request/limit* for a resource is the higher of:
  * the sum of all app containers request/limit for a resource
  * the effective init request/limit for a resource
* Scheduling is done based on effective requests/limits, which means
  init containers can reserve resources for initialization that are not used
  during the life of the Pod.
* The QoS (quality of service) tier of the Pod's *effective QoS tier* is the

>   QoS tier for init containers and app containers alike.

Reviewer 2: this is not true anymore.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers#exposing-pod-resource-requirements

Reviewer: Good spot!

Reviewer: @t-lakshmi, would you be comfortable revising this text solo? If you'd like to ask for help, please do that.

Author: @sftim, @gjkim42,
I revised the text and tried to fix it. PTAL once,
Thanks.

---
title: API for sidecar containers
content_type: concept
weight: 50
---

<!-- overview -->

> {{< feature-state for_k8s_version="v1.28" state="alpha" >}}

Reviewer: Looks right, thanks.

Sidecar containers are the secondary containers that runs along with the main
application container within the same {{< glossary_tooltip text="Pod" term_id="pod" >}}.
These containers are used to enhance or extend the functionality of the primary container
by providing additional services, or functionality such as logging, monitoring, security,
or data synchronization, without directly altering the primary application code.

<!-- body -->

Starting with Kubernetes 1.28 in alpha, a feature gate named `SidecarContainers`

> allows you to specify a `restartPolicy` for init containers which is independent of

Reviewer: ```suggestion
allows you to specify a `restartPolicy` for containers listed with a Pods `initContainers` field. These
restartable _sidecar_ containers are independent of
```
?

---

> title: API for sidecar containers

Reviewer 3: In the `concepts/workloads/pods` section, it is weird to focus on the API of a resource rather than the resource itself.

We may want to simplify to "Sidecar containers".

Author: @sftim,
Can you please share your thoughts on this.

Reviewer: Yes: @tengqm was right.

Author: Thanks, Reverted back to "Sidecar Containers".

content_type: concept

> weight: 50

Reviewer 3: Did we check if 50 is an appropriate value in this subdirectory?

---

<!-- overview -->
{{< feature-state for_k8s_version="v1.28" state="alpha" >}}

> Sidecar containers are the secondary containers that runs along with the main

Reviewer 3: ```suggestion
Sidecar containers are the secondary containers that run along with the main
```

Author: updated.

application container within the same {{< glossary_tooltip text="Pod" term_id="pod" >}}.

> These containers are used to enhance or extend the functionality of the primary container

Reviewer 3: Please **be consistent** when you are using a term. If you call something "main containers", or "application containers", then use it across the page. Do not introduce a term like "primary containers". This confuses the users.

```suggestion
These containers are used to enhance or to extend the functionality of the main application container
```

Author: Thanks for the suggestion, i will keep in mind.

by providing additional services, or functionality such as logging, monitoring, security,
or data synchronization, without directly altering the primary application code.

<!-- body -->

> Starting with Kubernetes 1.28 in alpha, a feature gate named `SidecarContainers`

Reviewer 3: This reads like "kubernetes 1.28 in alpha", ... so it is weird.

Author: updated.

> allows you to specify a `restartPolicy` for containers listed with a Pod's `initContainers`

Reviewer 3: ```suggestion
allows you to specify a `restartPolicy` for containers listed in a Pod's `initContainers`
```

Author: updated.

field. These restartable _sidecar_ containers are independent of other
[init containers](/docs/concepts/workloads/pods/init-containers/) and main application
container within the same pod. These can be started, stopped, or restarted independently

> of the main container and other init containers.

Reviewer 3: "of the ...." is wordy. If you are reading the whole sentence/paragraph, you will find that there are words that can be dropped. Dropping unnecessary words without sacrificing the readability is desirable.

Author: Updated the paragraph, please chack and let me know if still needs correction.

Container [probes](/docs/concepts/workloads/pods/pod-lifecycle/#types-of-probe)

> can also be added to control their lifecycle.

Reviewer 3: Avoid passive tense if possible.
We should instead say "You can add probes to a sidecar container...".

Author: updated.

If an init container is created with its `restartPolicy` set to `Always`, it will
start and remain running during the entire life of the Pod, which is useful for

> running supporting services separated from the main application containers.

Reviewer 3: Please avoid lengthy sentences like this.

Author: Updated.

If a `readinessProbe` is specified for this init container, its result will be used

> to determine the `ready` state of the Pod.

Reviewer 3: If you want to talk about "probes", then consolidate those words into one or two paragraphs. Please avoid jumping between (sub)topics back and forth.

be mixed with other init containers into complex Pod initialization flows.

Compared to regular init containers, sidecar-style init containers continue to
run and the next init container can begin starting once the kubelet has set
the `started` container status for the sidecar-style init container to true.
That status either becomes true because there is a process running in the
container and no startup probe defined, or
as a result of its `startupProbe` succeeding.

This feature can be used to implement the sidecar container pattern in a more
robust way, as the kubelet always restarts a sidecar container if it fails.

Here's an example of a Deployment with two containers, one of which is a sidecar:

{{% code_sample language="yaml" file="application/deployment-sidecar.yaml" %}}

This feature is also useful for running Jobs with sidecars, as the sidecar
container will not prevent the Job from completing after the main container
has finished.

Here's an example of a Job with two containers, one of which is a sidecar:

{{% code_sample language="yaml" file="application/job/job-sidecar.yaml" %}}
By default, this feature is not available in Kubernetes. To avail this feature,

> we need to enable the feature gate `SidecarContainers` by setting the value **true**

Reviewer 3: We always avoid "we".
You may want to reread the contribute guidelines about the content styles.

Author: Updated, Thanks.

---
title: Sidecar Containers
content_type: concept
weight: 50
---

<!-- overview -->
Sidecar containers are the secondary containers that runs along with the main
application container within the same {{< glossary_tooltip text="Pod" term_id="pod" >}}.
These containers are used to enhance or extend the functionality of the primary container
by providing additional services, or functionality such as logging, monitoring, security,
or data synchronization, without directly altering the primary application code.

<!-- body -->

## API for sidecar containers

> {{< feature-state for_k8s_version="v1.28" state="alpha" >}}

Reviewer: This should move right to the top of the page now it's a standalone concept.

Author: I was little confused on it and updated as per my understanding, PTAL and let me know if my understanding is wrong.

run and the next init container can begin starting once the kubelet has set
the `started` container status for the sidecar-style init container to true.
That status either becomes true because there is a process running in the
container and no startup probe defined, or
as a result of its `startupProbe` succeeding.

This feature can be used to implement the sidecar container pattern in a more
robust way, as the kubelet always restarts a sidecar container if it fails.

Here's an example of a Deployment with two containers, one of which is a sidecar:

{{% code_sample language="yaml" file="application/deployment-sidecar.yaml" %}}

This feature is also useful for running Jobs with sidecars, as the sidecar
container will not prevent the Job from completing after the main container
has finished.

Here's an example of a Job with two containers, one of which is a sidecar:

{{% code_sample language="yaml" file="application/job/job-sidecar.yaml" %}}
By default, this feature is not available in Kubernetes. To avail this feature,
we need to enable the feature gate `SidecarContainers` by setting the value **true**
to it.

> 

Reviewer: **_Optional_ feedback**

You could consider adding two new sections:

- “Differences from regular containers”
- “Differences from sidecar containers”

Author: > “Differences from sidecar containers”

Do you mean init containers?

Reviewer: Ah yes.

## Resource sharing within containers

Given the ordering and execution for init containers, the following rules
for resource usage apply:

* The highest of any particular resource request or limit defined on all init
  containers is the *effective init request/limit*. If any resource has no
  resource limit specified this is considered as the highest limit.
* The Pod's *effective request/limit* for a resource is the higher of:
  * the sum of all app containers request/limit for a resource
  * the effective init request/limit for a resource
* Scheduling is done based on effective requests/limits, which means
  init containers can reserve resources for initialization that are not used
  during the life of the Pod.
* The QoS (quality of service) tier of the Pod's *effective QoS tier* is the
  QoS tier for init containers and app containers alike.

Quota and limits are applied based on the effective Pod request and
limit.

Pod level control groups (cgroups) are based on the effective Pod request and
limit, the same as the scheduler.

## {{% heading "whatsnext" %}}

> * Read a blog post on [native sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/).

Reviewer 4: Instead of absolute link, we could use relative link here.
```suggestion
* Read a blog post on [native sidecar containers](/blog/2023/08/25/native-sidecar-containers/).
```

Author: Good catch!
Updated the same.

Reviewer 3: @T-Lakshmi Please click the "Resolve conversation" button when you fully agreed to the suggestions made by other reviwers AND you have updated the PR accordingly.
Marking conversation resolved saves a lot of time for other reviewers and yourself.
You know what are the discussion topics unsettled, what are the ones that have been resolved. A PR becomes cleaner on the GitHub webiste.
